compiler/x/codegen/BinaryEvaluator.cpp

```diff
@@ -1933,28 +1933,51 @@ TR::Register *OMR::X86::TreeEvaluator::integerDualMulEvaluator(TR::Node *node, T
 
 TR::Register *OMR::X86::TreeEvaluator::integerMulEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   TR::Register *targetRegister = 0;
-   TR::Node     *firstChild     = node->getFirstChild();
-   TR::Node     *secondChild    = node->getSecondChild();
-   bool          nodeIs64Bit    = TR::TreeEvaluator::getNodeIs64Bit(node, cg);
-   intptrj_t    value;
-
+   TR::Register *targetRegister  = 0;
+   TR::Node     *firstChild      = node->getFirstChild();
+   TR::Node     *secondChild     = node->getSecondChild();
+   bool          nodeIs64Bit     = TR::TreeEvaluator::getNodeIs64Bit(node, cg);
+   bool     oursIsTheOnlyMemRef  = true;
+   bool     countsAreDecremented = false;
+   bool firstChildRefCountIsZero = false;
+   TR::MemoryReference *tempMR   = NULL;
+   bool isMemRef                 = firstChild->getOpCode().isMemoryReference();
+   TR::DataType dataType         = secondChild->getDataType();
+   int64_t       value;
+   
    if (node->isDualCyclic())
       {
       return TR::TreeEvaluator::integerDualMulEvaluator(node, cg);
       }
+   
+   if(isMemRef)
+      {
+      // Make sure the original value is evaluated before the update if it
+      // is going to be used again.
+      if (firstChild->getReferenceCount() > 1)
```

Comment:
Can you explain why do we need this branch as well as `oursIsTheOnlyMemRef `? It seems when firstChild's reference count is larger than 2 we would always use the register version and a memory reference is not needed.